### PR TITLE
[timer] init timer + add CSR instructions

### PR DESCRIPTION
**Pull request by Bryanskiy**

### CODE REVIEW

File: include/common/state.hh [hunk 1]
```diff
@@ -44,8 +46,31 @@ public:
...
 
   RegVal get(CSRegId regnum) const { return regs.at(regnum); }
-
   void set(CSRegId regnum, RegVal val) { regs.at(regnum) = val; }
+
+  void updateTimers(OpType type) {
+    using CSRBindings = Counters::CSRBindings;
+
+    auto join = [](Word lower, Word upper) -> DWord {
+      auto tmp = static_cast<DWord>(upper) << (kBitsInByte * sizeof(Word));
```
Comment (derzhavin3016) [inline include/common/state.hh]:
```suggestion
      auto tmp = static_cast<DWord>(upper) << sizeofBits<Word>());
```

File: include/common/state.hh [hunk 2]
```diff
@@ -44,8 +46,31 @@ public:
...
+      return tmp | lower;
+    };
+
+    auto split = [](DWord res, Word &lower, Word &upper) {
+      constexpr auto separator = kBitsInByte * sizeof(Word);
+      constexpr auto end = kBitsInByte * sizeof(DWord);
```
Comment (derzhavin3016) [inline include/common/state.hh]:
```suggestion
      constexpr auto separator = sizeofBits<Word>();
      constexpr auto end = sizeofBits<DWord>();
```

File: include/common/counters.hh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        {OpType::SLTIU, 1},   {OpType::LUI, 2},   {OpType::AUIPC, 3},
+        {OpType::SLLI, 2},    {OpType::SRLI, 2},  {OpType::SRAI, 2},
+        {OpType::SLL, 2},     {OpType::SRL, 2},   {OpType::SRA, 2},
+        {OpType::BEQ, 36},    {OpType::BNE, 36},  {OpType::BLT, 36},
+        {OpType::BLTU, 36},   {OpType::BGEU, 36}, {OpType::BGE, 36},
+        {OpType::XOR, 1},
+    };
+
+    return throughput_.at(type);
+  }
```
Comment (derzhavin3016) [inline include/common/counters.hh]: I suppose you should provide default throughput if we've met an unknown instruction (maybe with some warn log message)

File: include/common/counters.hh [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  /* name bindings to CSR */
+  enum CSRBindings {
+    cycle = 0xC00,   /* Cycle counter for RDCYCLE instruction */
+    time = 0xC01,    /* Timer for RDTIME instruction */
+    instret = 0xC02, /* Instructions-retired counter for RDINSTRET */
+
+    cycleh = 0xC80,  /* Upper 32 bits of cycle, RV32 only */
+    timeh = 0xC81,   /* Upper 32 bits of time, RV32 only */
+    insreth = 0xC82, /* Upper 32 bits of instret, RV32 only */
+  };
```
Comment (derzhavin3016) [inline include/common/counters.hh]: Please, follow our codestyle

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

**Comment by Bryanskiy** [comment]
closes #40

File: include/common/counters.hh [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        {OpType::XOR, 1},     {OpType::CSRRW, 1},  {OpType::CSRRS, 1},
+        {OpType::CSRRC, 1},   {OpType::CSRRWI, 1}, {OpType::CSRRSI, 1},
+        {OpType::CSRRCI, 1},
+    };
+    auto it = throughput.find(type);
+    if (it == throughput.end()) {
+      spdlog::warn("Cant calculate {} throughput",
+                   static_cast<std::uint8_t>(type));
+      return 0;
+    }
```
Comment (derzhavin3016) [inline include/common/counters.hh]: Maybe return 1 here? Just to count this instruction

**Review by derzhavin3016 — CHANGES_REQUESTED** [review]

**Review by derzhavin3016 — APPROVED** [review]
LGBT

**Pull request merged**